Eliminate data_config min and max keys and SPLINTERDB_MAX_KEY_SIZE

rtjohnso opened the pull request:
Adds a `key` data type similar to the `message` type.

rosenhouse commented:
Hey @rtjohnso you've haven't requested a review on this yet, but I was curious and tried it out a bit this afternoon anyway.  I hope you don't mind!

I hit one issue that might be new to you: On this branch, when `data_config.key_size` is large (e.g. 103), but the application inserts tuples with small keys (e.g. 24 bytes) and large values (e.g. 1000 bytes) then after about 200,000 inserts, an assertion trips:
```
Assertion failed at src/trunk.c:5511:trunk_split_leaf(): "key_length(curr_key) == trunk_key_size(spl)".
```

I made a minimal unit-test here: https://github.com/vmware/splinterdb/compare/robj/key-datatype...gabe/test-variable-size-tuples-atop-robj-key-datatype

See it fail with 
```
make BUILD_MODE=debug && build/debug/bin/unit/splinterdb_key_size_test
```

The same unit test will pass on `main`.

Hope this is helpful!

rtjohnso commented:
Thanks for catching that.

The assertion is just wrong after my changes.  `trunk_key_size` is really
`trunk_max_key_size`, so the assertion should be
```

key_length(curr_key) <= trunk_max_key_size(spl)

```


I've made those changes (to the assertion and to rename
`trunk_key_size to `trunk_max_key_size`) and audited the code for any
other similar errors.


I also renamed `data_config.key_size` to `max_key_size` (and in
`master_config`).


I rebased your branch on those changes and your test now passes.  Thanks!


Best,

Rob




On Wed, Oct 26, 2022 at 5:17 PM Gabe Rosenhouse ***@***.***>
wrote:

> Hey @rtjohnso <https://github.com/rtjohnso> you've haven't requested a
> review on this yet, but I was curious and tried it out a bit this afternoon
> anyway. I hope you don't mind!
>
> I hit one issue that might be new to you: On this branch, when
> data_config.key_size is large (e.g. 103), but the application inserts
> tuples with small keys (e.g. 24 bytes) and large values (e.g. 1000 bytes)
> then after about 200,000 inserts, an assertion trips:
>
> Assertion failed at src/trunk.c:5511:trunk_split_leaf(): "key_length(curr_key) == trunk_key_size(spl)".
>
> I made a minimal unit-test here:
> robj/key-datatype...gabe/test-variable-size-tuples-atop-robj-key-datatype
> <https://github.com/vmware/splinterdb/compare/robj/key-datatype...gabe/test-variable-size-tuples-atop-robj-key-datatype>
>
> See it fail with
>
> make BUILD_MODE=debug && build/debug/bin/unit/splinterdb_key_size_test
>
> The same unit test will pass on main.
>
> Hope this is helpful!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/splinterdb/pull/464#issuecomment-1292801394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA45FRPH6Z7UGC5PA47TMFDWFHCZ7ANCNFSM6AAAAAARNSD2JU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rtjohnso commented:
This extends and includes PR #460  `Robj/trunk variable sized keys`, so you can review them as one big PR now or do #460 first and then do the delta on this one after #460 is done.

ajhconway commented on tests/test_util.h:
```diff
@@ -0,0 +1,27 @@
+// Copyright 2018-2021 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef __TEST_UTIL_H__
+#define __TEST_UTIL_H__
+
+#include "data_internal.h"
+
+static inline bool
+keys_equal_contents(key a, key b)
+{
+   return slice_lex_cmp(a.user_slice, b.user_slice) == 0;
+}
+
+static inline key
+writable_buffer_to_key(writable_buffer *wb)
+{
+   return key_create_from_slice(writable_buffer_to_slice(wb));
+}
+
+static inline void
+writable_buffer_copy_key(writable_buffer *wb, key src)
+{
+   writable_buffer_copy_slice(wb, key_slice(src));
+}
+
+#endif /* __TEST_UTIL_H__ */
```
I guess I feel like these aren't really test specific, so I'd put them in regular `util.h`. I also don't understand why we don't just use `keys_equal` in the tests?

rtjohnso replied:
I replaced all uses of writable_buffers for keys with key_buffers, so these all went away.  File deleted.

ajhconway commented on tests/functional/log_test.c:
```diff
@@ -128,17 +140,19 @@ test_log_thread(void *arg)
       log_write(logh, skey, merge_accumulator_to_message(&msg), i);
    }
+
+   merge_accumulator_deinit(&msg);
```
This is just an incidental bugfix?

rtjohnso replied:
yes

ajhconway commented on src/data_internal.h:
```diff
@@ -160,29 +160,140 @@ merge_accumulator_is_null(const merge_accumulator *ma)
+static inline key
+key_create_from_slice(slice user_slice)
+{
+   return (key){.kind = user_key, .user_slice = user_slice};
```
Is there some documentation about how keys/messages/slices work and what the assumptions/limitations are?

rtjohnso replied:
Wrote some.

ajhconway reviewed (changes requested):
This is a great change and I'm happy with a lot here.

This is also a huge change and I'm not even confident that I've reviewed this enough.

There's tons of minor stuff. Most notably lots of `key key` definitions and stuff like that.

The big change I want is I think we should standardize a on-disk key format for the mini allocator and the trunk and just use that, rather than the weird length encoding in the mini_allocator and the ad hoc length/kind encoding in the pivot_data.

rtjohnso commented on src/btree.c:
```diff
@@ -1512,7 +1499,7 @@ btree_defragment_or_split_child_index(cache              *cc,
                                       uint64      index_of_child_in_parent,
                                       btree_node *child,
-                                      const slice key_to_be_inserted,
+                                      key         key_to_be_inserted,
```
`const` is kinda meaningless on slices passed-by-value, since the pointer within the slice is already const.  It was my mistake to put const in in the first place.  So I've just been removing them as I go.

rtjohnso commented on src/btree.c:
```diff
@@ -151,7 +148,7 @@ void
 {
    for (int i = 0; i < hdr->num_entries; i++) {
-      slice key = btree_get_tuple_key(cfg, hdr, i);
+      key key = btree_get_tuple_key(cfg, hdr, i);
```
All fixed

rtjohnso commented on src/data_internal.c:
```diff
@@ -1,6 +1,14 @@
 #include "data_internal.h"
 
+key NEGATIVE_INFINITY_KEY = {.kind       = negative_infinity,
+                             .user_slice = {(uint64)-1, NULL}};
```
The problem was that `NULL_SLICE` was defined as an `extern const slice`, which means it cannot be used in the initializers of other global constants.

I changed all these global constants to be `#define`s of initializers.

rtjohnso commented on src/data_internal.c:
```diff
@@ -1,6 +1,14 @@
+key POSITIVE_INFINITY_KEY = {.kind       = positive_infinity,
+                             .user_slice = {(uint64)-1, NULL}};
+
+key NULL_KEY = {.kind = user_key, .user_slice = {0, NULL}};
```
See above

rtjohnso commented on src/data_internal.h:
```diff
@@ -160,29 +160,140 @@ merge_accumulator_is_null(const merge_accumulator *ma)
 }
 
+typedef enum {
+   positive_infinity = 1,
+   user_key          = 2,
+   negative_infinity = 3
```
done

rtjohnso commented on src/data_internal.h:
```diff
@@ -160,29 +160,140 @@ merge_accumulator_is_null(const merge_accumulator *ma)
+} key_type;
+
+typedef struct key {
+   key_type kind;
```
I don't think it will matter since this is an in-memory structure and alignment will force padding anyway.  But yes we should do that on disk.

rtjohnso commented on src/data_internal.h:
```diff
@@ -160,29 +160,140 @@ merge_accumulator_is_null(const merge_accumulator *ma)
+}
+
+static inline key
+data_min_key(const data_config *cfg)
```
done

rtjohnso commented on src/data_internal.h:
```diff
@@ -160,29 +160,140 @@ merge_accumulator_is_null(const merge_accumulator *ma)
+data_key_compare(const data_config *cfg, key key1, key key2)
 {
-   return cfg->key_compare(cfg, key1, key2);
+   if (key_is_negative_infinity(key1)) {
```
nice. done.

rtjohnso commented on src/splinterdb.c:
```diff
@@ -845,9 +584,9 @@ struct splinterdb_iterator {
-                         slice                 start_key // IN
+splinterdb_iterator_init(const splinterdb     *kvs,           // IN
+                         splinterdb_iterator **iter,          // OUT
+                         slice                 user_start_key // IN
```
`key` is not currently part of the external api, so users still use NULL_SLICEs for +/- infinity.

rtjohnso commented on src/mini_allocator.c:
```diff
@@ -75,19 +80,31 @@ typedef struct ONDISK unkeyed_meta_entry {
 sizeof_keyed_meta_entry(const keyed_meta_entry *entry)
 {
-   return sizeof(keyed_meta_entry) + entry->start_key_length;
+   if (entry->start_key_length == POSITIVE_INFINITY_KEY_LENGTH
+       || entry->start_key_length == NEGATIVE_INFINITY_KEY_LENGTH)
+   {
+      return sizeof(keyed_meta_entry);
+   } else {
+      return sizeof(keyed_meta_entry) + entry->start_key_length;
+   }
```
0 is a valid user key length.

rtjohnso commented on src/shard_log.c:
```diff
@@ -222,8 +223,10 @@ get_new_page_for_thread(shard_log             *log,
-shard_log_write(log_handle *logh, slice key, message msg, uint64 generation)
+shard_log_write(log_handle *logh, key key, message msg, uint64 generation)
 {
+   debug_assert(key_is_user_key(key));
```
That's a reasonable question.  However, the general reason I want to use keys everywhere is for future support for huge keys (i.e. blobs).

When blob support for huge keys lands, the flag indicating whether the key is inline or a blob will be stored in the key structure.  The slice in the key will hold either the inline key or the blob descriptor (i.e. the array of pointers to the extents and pages used by the blob).  And this function will need to know which kind of key it is dealing with.  So it will need a key, not a slice.

rtjohnso commented on src/trunk.c:
```diff
@@ -1128,40 +1133,66 @@ trunk_subtract_subbundle_filter_number(trunk_handle *spl,
    // debug asserts (should be optimized away)
    if (pivot_no != 0) {
-      __attribute__((unused)) const char *pred_pivot =
+      __attribute__((unused)) key pred_pivot =
```
Audited the code and converted all naked `__attribute__((unused)) to `debug_only`, except a few that were actually used!

rtjohnso commented on tests/unit/splinterdb_quick_test.c:
```diff
@@ -1004,7 +944,7 @@ check_current_tuple(splinterdb_iterator *it, const int expected_i)
    int rc = 0;
 
-   char expected_key[MAX_KEY_SIZE]        = {0};
+   char expected_key[7]                   = {0};
```
cleaned up

rtjohnso commented on src/data_internal.h:
```diff
@@ -160,29 +160,140 @@ merge_accumulator_is_null(const merge_accumulator *ma)
+keys_equal(key a, key b)
+{
+   return a.kind == b.kind
+          && IMPLIES(a.kind == user_key,
+                     slices_equal(a.user_slice, b.user_slice));
```
Changed to blank_equals for both slices and keys.

rtjohnso commented on src/mini_allocator.c:
```diff
@@ -60,6 +60,11 @@ typedef struct ONDISK keyed_meta_entry {
 } keyed_meta_entry;
 
+#define POSITIVE_INFINITY_KEY_LENGTH                                           \
+   ((FTYPEOF(keyed_meta_entry, start_key_length)) - 1)
```
Eliminated FTYPEOF and FSIZEOF as part of generalizing the on-disk encodings of keys and tuples.

rtjohnso commented on src/platform_linux/platform.h:
```diff
@@ -61,6 +61,8 @@
  */
 #define ARRAY_SIZE(x) ASSERT_EXPR(IS_ARRAY(x), (sizeof(x) / sizeof((x)[0])))
 
+#define FTYPEOF(s, f) typeof(((typeof(s) *)NULL)->f)
```
Gone

rtjohnso commented on src/trunk.c:
```diff
@@ -567,8 +568,11 @@ typedef struct ONDISK trunk_pivot_data {
+   uint16         key_length;
 } trunk_pivot_data;
 
+#define INFINITY_KEY_LENGTH ((FTYPEOF(trunk_pivot_data, key_length)) - 1)
```
Done

rtjohnso commented on src/trunk.c:
```diff
@@ -1170,31 +1201,45 @@ trunk_set_pivot(trunk_handle *spl,
                          page_handle  *node,
-                         const char   *min_key,
-                         const char   *max_key)
+                         key           min_key,
+                         key           max_key)
```
Done

rtjohnso commented on src/trunk.c:
```diff
@@ -1170,31 +1201,45 @@ trunk_set_pivot(trunk_handle *spl,
    pdata->srq_idx = -1;
-   dst_pivot_key  = trunk_get_pivot(spl, node, 1);
-   memmove(dst_pivot_key, max_key, trunk_key_size(spl));
+   if (key_is_negative_infinity(min_key)) {
```
That would be a hassle because `trunk_set_pivot` checks the sortedness of the pivots in debug mode, which would fail when setting the first pivot (since the second hasn't been set).

To avoid this, we'd have to set `hdr->num_pivot_keys` to 1, then set the first pivot, then set the `num_pivot_keys` to 2, then set the second pivot.  Seems like unnecessary hassle to me.

rtjohnso commented on src/trunk.c:
```diff
@@ -1458,8 +1502,8 @@ trunk_shift_pivots(trunk_handle *spl,
 
-   char  *dst_pivot       = trunk_get_pivot(spl, node, pivot_no + shift);
-   char  *src_pivot       = trunk_get_pivot(spl, node, pivot_no);
+   char  *dst_pivot       = trunk_get_pivot_buffer(spl, node, pivot_no + shift);
+   char  *src_pivot       = trunk_get_pivot_buffer(spl, node, pivot_no);
```
Now uses standardized key encoding.

rtjohnso commented on src/trunk.c:
```diff
@@ -1502,12 +1547,12 @@ trunk_add_pivot_new_root(trunk_handle *spl,
-   const char *pivot_key                       = trunk_get_pivot(spl, child, 0);
-   __attribute__((unused)) const char *min_key = spl->cfg.data_cfg->min_key;
+   key                               pivot_key = trunk_get_pivot(spl, child, 0);
+   __attribute__((unused)) const key min_key = data_min_key(spl->cfg.data_cfg);
```
All this debug code is pointless in the new system and is gone.

rtjohnso commented on src/trunk.c:
```diff
@@ -1787,18 +1828,61 @@ trunk_pivot_tuples_in_branch_slow(trunk_handle *spl,
+
+static inline key
+key_buffer_key(key_buffer *kb)
+{
+   if (kb->kind == negative_infinity) {
+      return NEGATIVE_INFINITY_KEY;
+   } else if (kb->kind == positive_infinity) {
+      return POSITIVE_INFINITY_KEY;
+   } else {
+      return key_create_from_slice(writable_buffer_to_slice(&kb->wb));
+   }
+}
+
+static inline key
+key_buffer_init_from_key(key_buffer *kb, platform_heap_id hid, key key)
+{
+   if (key_is_negative_infinity(key)) {
+      kb->kind = negative_infinity;
+   } else if (key_is_positive_infinity(key)) {
+      kb->kind = positive_infinity;
+   } else {
+      kb->kind = user_key;
+      writable_buffer_init_with_buffer(
+         &kb->wb, hid, sizeof(kb->default_buffer), kb->default_buffer, 0);
+      writable_buffer_copy_slice(&kb->wb, key_slice(key));
+   }
+   return key_buffer_key(kb);
+}
+
+static inline void
+key_buffer_deinit(key_buffer *kb)
+{
+   if (kb->kind == user_key) {
+      writable_buffer_deinit(&kb->wb);
+   }
+}
+
+#define KEY_CREATE_LOCAL_COPY(dst, hid, src)                                   \
+   key_buffer dst##kb;                                                         \
+   key        dst = key_buffer_init_from_key(&dst##kb, hid, src)
```
Moved to `data_internal.h`

rtjohnso commented on src/trunk.h:
```diff
@@ -219,6 +219,20 @@ struct trunk_handle {
    trunk_compacted_memtable compacted_memtable[/*cfg.mt_cfg.max_memtables*/];
 };
 
+/*
```
done

rtjohnso commented on src/util.h:
```diff
@@ -181,6 +181,25 @@ writable_buffer_init(writable_buffer *wb, platform_heap_id heap_id)
       wb, heap_id, 0, NULL, WRITABLE_BUFFER_NULL_LENGTH);
 }
 
+/*
```
In `data_internal.h`

rtjohnso commented on src/util.h:
```diff
@@ -238,6 +257,15 @@ writable_buffer_append(writable_buffer *wb, uint64 length, const void *newdata)
 }
 
+/*
+ * Creates a copy of src in newly declared slice dst.  Everything is
```
yes

rtjohnso commented on test.sh:
```diff
@@ -560,7 +560,7 @@ function run_splinter_functionality_tests() {
                                             --seed "$SEED"
 
-    max_key_size=105
+    max_key_size=103
```
Not available in the shell script?

rtjohnso commented on tests/functional/btree_test.c:
```diff
@@ -488,10 +488,10 @@ test_btree_async_lookup(cache                   *cc,
    btree_ctxt_init(
       &async_ctxt->ctxt, &async_ctxt->cache_ctxt, btree_test_async_callback);
-   slice key_slice = writable_buffer_to_slice(&async_ctxt->key);
+   key key_key = writable_buffer_to_key(&async_ctxt->key);
```
all fixed

rtjohnso commented on src/trunk.c:
```diff
@@ -5990,6 +6081,13 @@ trunk_range_iterator_deinit(trunk_range_iterator *range_itor)
    }
+
+   key_buffer_deinit(&range_itor->min_key);
+   if (range_itor->has_max_key) {
```
done

rtjohnso commented on src/trunk.c:
```diff
@@ -5696,33 +5772,36 @@ const static iterator_ops trunk_range_iterator_ops = {
    range_itor->num_tuples   = num_tuples;
-   if (min_key == NULL) {
-      min_key = spl->cfg.data_cfg->min_key;
+   if (key_is_null(min_key)) {
```
done

rtjohnso commented on src/data_internal.h:
```diff
@@ -160,29 +160,140 @@ merge_accumulator_is_null(const merge_accumulator *ma)
+static inline key
+key_create(uint64 length, const void *data)
+{
+   return (key){.kind = user_key, .user_slice = slice_create(length, data)};
```
yes.  See NULL_KEY, eg.

rtjohnso commented on tests/functional/filter_test.c:
```diff
@@ -169,14 +171,16 @@ test_filter_perf(cache           *cc,
-         ZERO_ARRAY(key);
-         *(uint64 *)key = idx;
-         fp_arr[idx]    = cfg->hash(key, key_size, cfg->seed);
+         memset(keybuf, 0, key_size);
```
done

rtjohnso commented on tests/functional/btree_test.c:
```diff
@@ -185,11 +187,10 @@ test_btree_insert_thread(void *arg)
-   uint64            start_time = platform_get_timestamp();
-   writable_buffer   key;
+   uint64 start_time = platform_get_timestamp();
+   WRITABLE_BUFFER_DEFAULT(key, NULL);
```
Audited and fixed all of these.  (Also for merge_accumulator_init)

rtjohnso commented on tests/functional/btree_test.c:
```diff
@@ -1012,12 +998,10 @@ test_btree_merge_basic(cache             *cc,
-                    ? NULL_SLICE
-                    : slice_create(btree_cfg->data_cfg->key_size,
-                                   pivot_key[pivot_no + 1].k);
+      key lo = writable_buffer_to_key(&pivot_key[pivot_no]);
+      key hi = pivot_no == arity - 1
+                  ? POSITIVE_INFINITY_KEY
+                  : writable_buffer_to_key(&pivot_key[pivot_no + 1]);
```
Done throughout the code.

rtjohnso commented on src/trunk.c:
```diff
@@ -7984,40 +8059,35 @@ trunk_print_pivots(platform_log_handle *log_handle,
-                      (int)sizeof(key_string),
-                      key_string);
+      }
+      if (key_is_user_key(pivot)) {
```
I plan to do an audit to convert all printing to use the `key_string` and `message_string` macros.  But I want to do it as a separate PR, since this one is already huge.

Also, I'm not sure what full key-printing function we have, if any.
